server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -2464,203 +2465,255 @@ public boolean assertAllListenersResolved() {
      * Package private to allow for tests.
      */
     static final ClusterStateTaskExecutor<ShardSnapshotUpdate> SHARD_STATE_EXECUTOR = (currentState, tasks) -> {
-        int changedCount = 0;
-        int startedCount = 0;
+        final SnapshotShardsUpdateContext update = new SnapshotShardsUpdateContext(currentState, tasks);
+        for (SnapshotsInProgress.Entry entry : currentState.custom(SnapshotsInProgress.TYPE, SnapshotsInProgress.EMPTY).entries()) {
```

Comment:
It seems strange to me to rely on the caller to apply it to every entry in the state. IMO this loop should be inside the new object (or else the new object should return the updated version of each entry).

Reply:
++ moved the loop into the new object now